# FIT5197 2023 S1 Final Assessment

**SPECIAL NOTE:** Please refer to the [assessment page](https://lms.monash.edu/mod/assign/view.php?id=11961926) for rules, general guidelines and marking rubrics of the assessment (the marking rubric for the kaggle competition part will be released near the deadline in the same page). Failure to comply with the provided information will result in a deduction of mark (i.e. late penalties) or breach of academic integrity.

**YOUR NAME**: Gautam Ravi Kumar

**STUDENT ID**: 33197970

**KAGGLE NAME/ID** (See part 1, Question 5 or part 2): Gautam Ravi Kumar

Please also enter your details in this [google form](https://forms.gle/XZEck2BPKG2PftdC9).

# Part 1 Regression (50 Marks)

Around 1000 people were questioned in a [life and wellbeing survey](https://www.get-happier.com/) to build a model to predict happiness of an individual. You need to build regression models to optimally predict the variable in the survey dataset called 'happiness' based on any, or all, of the other survey question responses. 

You have been provided with two datasets, ```regression_train.csv``` and ```regression_test.csv```. Using these datasets, you hope to build a model that can predict happiness level using the other variables. ```regression_train.csv``` comes with the ground-truth target label (i.e. happiness level) whereas `regression_test.csv` comes with independent variables (input information) only.

On the order of around 70 survey questions have been converted into predictor variables that can be used to predict happiness. We do not list all the predictor names here, but their names given in the data header can clearly be linked to the survey questions. E.g. the predictor variable 'iDontFeelParticularlyPleasedWithTheWayIAm' corresponds to the survey question 'I don’t feel particularly pleased with the way I am.'

**PLEASE NOTE THAT THE USE OF LIBRARIES ARE PROHIBITED IN THESE QUESTIONS UNLESS STATED OTHERWISE, ANSWERS USING LIBRARIES WILL RECEIVE 0 MARKS**

## Question 1 (NO LIBRARIES ALLOWED) (4 Mark)
Please load the ```regression_train.csv``` and fit a [$\textbf{multiple linear regression model}$](https://en.wikipedia.org/wiki/Linear_regression) with 'happiness' being the target variable. According to the summary table, which predictors do you think are possibly associated with the target variable (use the significance level of 0.01), and which are the **Top 5** strongest predictors? Please write an R script to automatically fetch and print this information.

**NOTE**: Manually doing the above tasks will result in 0 marks.

In [1]:
# ANSWER BLOCK
# load the csv file
df <- read.csv("regression_train.csv")
# fit the multilinear regression model
fit <- lm(happiness ~ ., df)
summary(fit)


Call:
lm(formula = happiness ~ ., data = df)

Residuals:
     Min       1Q   Median       3Q      Max 
-21.9885  -5.2818  -0.0567   5.0144  21.1118 

Coefficients:
                                                                                             Estimate
(Intercept)                                                                                 -5.538505
genderMale                                                                                  -0.434500
alwaysDepressed                                                                              0.136364
alwaysAnxious                                                                               -0.080143
alwaysStressed                                                                              -0.748498
alwaysLoveAndCareForYourself                                                                 0.854401
alwaysAccountableAndResponsibleForYourActions                                                2.484598
alwaysCalm         

In [2]:
# Find the top 5 strongest predictors by p value
p_values <- head(sort(summary(fit)$coefficients[, "Pr(>|t|)"]), n = 5)
p_values <- names(p_values)
print("The top 5 predictors are:")
print(p_values)

[1] "The top 5 predictors are:"
[1] "alwaysAccountableAndResponsibleForYourActions"                           
[2] "iUsuallyHaveAGoodInfluenceOnEvents"                                      
[3] "doYouFeelComfortableEngagingInConversationsWithPeopleFromDifferSometimes"
[4] "betterDecisionsAreMadeWhenLeadersDiscussMattersWithOthersFirst"          
[5] "iAmAlwaysCommittedAndInvolved"                                           


## Question 2 (2 Mark)
[**R squared**](https://en.wikipedia.org/wiki/Coefficient_of_determination) from the summary table reflects that the full model doesn't fit the training dataset well; thus, you try to quantify the error between the values of the ground-truth and those of the model prediction. You want to write a function to predict 'happiness' with the given dataset and calculate the [root mean squared error (rMSE)](https://en.wikipedia.org/wiki/Root-mean-square_deviation) between the model predictions and the ground truths. Please test this function on the full model and the training dataset.

In [3]:
# ANSWER BLOCK
# Function to calculate rmse of a model, using actual values and predicted values
rmse_func <- function(true, predicted){
    return (sqrt(mean((true - predicted)^2)))
}
# Find predicted values
predictions <- predict(fit, newdata = df)
print(rmse_func(df$happiness,predictions))

[1] 7.492705


## Question 3 (2 Marks)
You find the full model complicated and try to reduce the complexity by performing [bidirectional stepwise regression](https://en.wikipedia.org/wiki/Stepwise_regression) with [BIC](https://en.wikipedia.org/wiki/Bayesian_information_criterion).

Calculate the **rMSE** of this new model with the function that you implemented previously. Is there anything you find unusual? Explain your findings in 100 words.

In [4]:
ncol(df)

[1] 86

In [5]:
# ANSWER BLOCK
# fit the  bidirectional stepwise regression model with BIC
back.fit = step(fit,k = log(nrow(df)), direction = "both")
# Find predicted values of the step model
predictions <- predict(back.fit, newdata = df)
summary(back.fit)
# find rsme
print(rmse_func(df$happiness,predictions))

Start:  AIC=3938.65
happiness ~ gender + alwaysDepressed + alwaysAnxious + alwaysStressed + 
    alwaysLoveAndCareForYourself + alwaysAccountableAndResponsibleForYourActions + 
    alwaysCalm + extreAustraliayGoodAbilityToSense + alwaysHaveDigestiveProblems + 
    myBodyIsHypermobileAndLovesToMove + alwaysHaveFun + alwaysSerious + 
    extreAustraliayGoodCommunicator + iDontFeelParticularlyPleasedWithTheWayIAm + 
    iAmIntenselyInterestedInOtherPeople + iFeelThatLifeIsVeryRewarding + 
    iHaveVeryWarmFeelingsTowardsAlmostEveryone + iRarelyWakeUpFeelingRested + 
    iAmNotParticularlyOptimisticAboutTheFuture + iFindMostThingsAmusing + 
    iAmAlwaysCommittedAndInvolved + iDoNotThinkThatTheWorldIsAGoodPlace + 
    iAmWellSatisfiedAboutEverythingInMyLife + iDontThinkILookAttractive + 
    thereIsAGapBetweenWhatIWouldLikeToDoAndWhatIHaveDone + iAlwaysHaveACheerfulEffectOnOthers + 
    iCanFitInFindTimeForEverythingIWantTo + iFeelThatIAmNotEspeciallyInControlOfMyLife + 
    iFeelAbleToTak


Call:
lm(formula = happiness ~ alwaysAccountableAndResponsibleForYourActions + 
    alwaysCalm + alwaysHaveDigestiveProblems + alwaysHaveFun + 
    iDontFeelParticularlyPleasedWithTheWayIAm + iRarelyWakeUpFeelingRested + 
    iFindMostThingsAmusing + iAmAlwaysCommittedAndInvolved + 
    thereIsAGapBetweenWhatIWouldLikeToDoAndWhatIHaveDone + iCanFitInFindTimeForEverythingIWantTo + 
    iFeelFullyMentallyAlert + iUsuallyHaveAGoodInfluenceOnEvents + 
    iDontHaveParticularlyHappyMemoriesOfThePast + onTheWholeIAmSatisfiedWithMyself + 
    iFeelThatIHaveANumberOfGoodQualities + iFeelThatImAPersonOfWorth + 
    betterDecisionsAreMadeWhenLeadersDiscussMattersWithOthersFirst + 
    involvementInJointProjectsIsVerySatisfying + committeesAreAWasteOfTime + 
    iLikeToGetToKnowPeopleBetter + friendsOughtToBeConcernedAboutEachOthersWishes + 
    doYouFeelComfortableEngagingInConversationsWithPeopleFromDiffer + 
    iAmOpenToEngagingInPracticesWhichFocusOnMyInterestsAndBeliefs + 
    iWillinglySe

[1] 8.142875


Despite the reduced complexity, the rMSE of the stepwise regression model(8.14) is greater than the regular linear model(7.49), implying that it performs slightly worse than the original model. The step function reduces the 86 features to a mere 28, and despite this, has only a 0.7 difference in rMSE. the reduction in performance is likely due to some of the features removed that may have been important in certain cases but not applicable in others.

## Question 4 (2 Mark)
Although stepwise regression has reduced the model complexity significantly, the model still contains a lot of variables that we want to remove. Therefore, you are interested in lightweight linear regression models with ONLY TWO predictors. Write a script to automatically find the best lightweight model which corresponds to the model with the least **rMSE** on the training dataset. Compare the **rMSE** of the best lightweight model with the **rMSE** of the full model - ```lm.fit``` - that you built previously. Give an explanation for these results based on consideration of the predictors involved.

In [6]:
# ANSWER BLOCK
# Find the 2 best predictors by p-value
cols <- head(sort(summary(back.fit)$coefficients[, "Pr(>|t|)"]),n = 2)
# Find the names of those predictors
cols <- names(cols)
# Get a subset of the data with only the predictors and the label
new_df <- subset(df, select = c(cols, 'happiness'))
# fit the linear regresson model
new_fit <- lm(happiness ~ ., new_df)
summary(new_fit)


Call:
lm(formula = happiness ~ ., data = new_df)

Residuals:
    Min      1Q  Median      3Q     Max 
-34.192  -6.312   0.106   6.656  30.447 

Coefficients:
                                              Estimate Std. Error t value
(Intercept)                                    -4.6689     0.4970  -9.394
alwaysAccountableAndResponsibleForYourActions   2.6692     0.3821   6.986
onTheWholeIAmSatisfiedWithMyself                4.6089     0.4265  10.807
                                              Pr(>|t|)    
(Intercept)                                    < 2e-16 ***
alwaysAccountableAndResponsibleForYourActions 5.89e-12 ***
onTheWholeIAmSatisfiedWithMyself               < 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 9.924 on 815 degrees of freedom
Multiple R-squared:  0.1894,	Adjusted R-squared:  0.1874 
F-statistic: 95.22 on 2 and 815 DF,  p-value: < 2.2e-16


In [7]:
# Find predicted values of the model
new_predictions <- predict(new_fit, newdata = new_df)
# Find rsme
rmse_func(df$happiness,new_predictions)

[1] 9.906075

The number of predictors had been further reduced to 2. This is done by selecting the 2 predictors with the lowest p values. However, the rSME of this new model is higher than the original model, likely due to too many important features being removed.

### ANSWER (TEXT)

## Question 5 (Libraries are allowed) (40 Marks)
As a Data Scientist, one of the key tasks is to build models $\textbf{most appropriate/closest}$ to the truth; thus, modelling will not be limited to the aforementioned steps in this assignment. To simulate for a realistic modelling process, this question will be in the form of a [Kaggle competition](https://www.kaggle.com/t/39b1b7b305ca4feaa234135059b701ac) among students to find out who has the best model.

Thus, you **will be graded** by the **rMSE** performance of your model, the better your model, the higher your score. Additionally, you need to describe/document your thought process in this model building process, this is akin to showing your working properly for the mathematic sections. If you don't clearly document the reasonings behind the model you use, we will have to make some deductions on your scores.

This is the [video tutorial](https://www.youtube.com/watch?v=rkXc25Uvyl4) on how to join any Kaggle competition. 

When you optimize your model's performance, you can use any supervised model that you know and feature selection might be a big help as well. [Check the non-exhaustive set of R functions relevant to this unit](https://lms.monash.edu/mod/resource/view.php?id=11405864) for ideas for different models to try.

$\textbf{Note}$ Please make sure that we can install the libraries that you use in this part, the code structure can be:

```install.packages("some package", repos='http://cran.us.r-project.org')```

```library("some package")```

Remember that if we cannot run your code, we will have to give you a deduction. Our suggestion is for you to use the standard ```R version 3.6.1```

You also need to name your final model ``fin.mod`` so we can run a check to find out your performance. A good test for your understanding would be to set the previous $\textbf{BIC model}$ to be the final model to check if your code works perfectly.

In [8]:
# Build your final model here, use additional coding blocks if you need to
# fit linear model
fit <- lm(happiness ~ ., df)
# uses step function with backward direction
fin.mod <- step(fit,k = 2, trace = 1, direction = "backward")

Start:  AIC=3486.79
happiness ~ gender + alwaysDepressed + alwaysAnxious + alwaysStressed + 
    alwaysLoveAndCareForYourself + alwaysAccountableAndResponsibleForYourActions + 
    alwaysCalm + extreAustraliayGoodAbilityToSense + alwaysHaveDigestiveProblems + 
    myBodyIsHypermobileAndLovesToMove + alwaysHaveFun + alwaysSerious + 
    extreAustraliayGoodCommunicator + iDontFeelParticularlyPleasedWithTheWayIAm + 
    iAmIntenselyInterestedInOtherPeople + iFeelThatLifeIsVeryRewarding + 
    iHaveVeryWarmFeelingsTowardsAlmostEveryone + iRarelyWakeUpFeelingRested + 
    iAmNotParticularlyOptimisticAboutTheFuture + iFindMostThingsAmusing + 
    iAmAlwaysCommittedAndInvolved + iDoNotThinkThatTheWorldIsAGoodPlace + 
    iAmWellSatisfiedAboutEverythingInMyLife + iDontThinkILookAttractive + 
    thereIsAGapBetweenWhatIWouldLikeToDoAndWhatIHaveDone + iAlwaysHaveACheerfulEffectOnOthers + 
    iCanFitInFindTimeForEverythingIWantTo + iFeelThatIAmNotEspeciallyInControlOfMyLife + 
    iFeelAbleToTak

In [9]:
# Load in the test data.
test <- read.csv("regression_test.csv")
# If you are using any packages that perform the prediction differently, please change this line of code accordingly.
pred.label <- predict(fin.mod, test)
# put these predicted labels in a csv file that you can use to commit to the Kaggle Leaderboard
write.csv(
    data.frame("RowIndex" = seq(1, length(pred.label)), "Prediction" = pred.label),  
    "RegressionPredictLabel.csv", 
    row.names = F
)

In [ ]:
## PLEASE DO NOT ALTER THIS CODE BLOCK, YOU ARE REQUIRED TO HAVE THIS CODE BLOCK IN YOUR JUPYTER NOTEBOOK SUBMISSION
## Please skip (don't run) this if you are a student
## For teaching team use only

tryCatch(
    {
        source("../supplimentary.R")
    },
    error = function(e){
        source("supplimentary.R")
    }
)

truths <- tryCatch(
    {
        read.csv("../regression_test_label.csv")
    },
    error = function(e){
        read.csv("regression_test_label.csv")
    }
)


RMSE.fin <- rmse(pred.label, truths$x)
cat(paste("RMSE is", RMSE.fin))

# Part 2 Classification (50 Marks)

Around 1000 people were questioned in a [life and wellbeing survey](https://www.get-happier.com/) to build a model to predict happiness of an individual, but this time we want to predict a categorical score for happiness, rather than a continuous score. You need to build 5-class classification models to optimally predict the variable in the survey dataset called 'iAmVeryHappy' based on any, or all, of the other survey question responses. 

You have been provided with two datasets, ```classification_train.csv``` and ```classification_test.csv```. Using these datasets, you hope to build a model that can predict 'iAmVeryHappy' using the other variables. ```classification_train.csv``` comes with the ground-truth target label (i.e. 'iAmVeryHappy' happiness classes) whereas `classification_test.csv` comes with independent variables (input information) only.

On the order of around 70 survey questions have been converted into predictor variables that can be used to predict 'iAmVeryHappy'. We do not list all the predictor names here, but their names given in the data header can clearly be linked to the survey questions. E.g. the predictor variable 'iDontFeelParticularlyPleasedWithTheWayIAm' corresponds to the survey question 'I don’t feel particularly pleased with the way I am.'

This question will also be in the form of a [Kaggle competition](https://www.kaggle.com/t/ab17f8b7e1d941b6bc70ce0265ca8798) among students to find out who has the best model.

In [10]:
# Import naive bayes
library(e1071)

Warning message:
"package 'e1071' was built under R version 3.6.3"

In [11]:
# Load in the train and test classification data.
train <- read.csv("classification_train.csv")
test <- read.csv("classification_test.csv")

# Build your final model here, use additional coding blocks if you need to

# fit a naive bayes model
fin.mod <- naiveBayes(as.factor(iAmVeryHappy) ~.,train)
# If you are using any packages that perform the prediction differently, please change this line of code accordingly.

# Added type = "class" to ensure that it is a classification model
pred.label <- predict(fin.mod, test, type = "class")
# put these predicted labels in a csv file that you can use to commit to the Kaggle Leaderboard
write.csv(
    data.frame("RowIndex" = seq(1, length(pred.label)), "Prediction" = pred.label),  
    "ClassificationPredictLabel.csv", 
    row.names = F
)

In [ ]:
## PLEASE DO NOT ALTER THIS CODE BLOCK, YOU ARE REQUIRED TO HAVE THIS CODE BLOCK IN YOUR JUPYTER NOTEBOOK SUBMISSION
## Please skip (don't run) this if you are a student
## For teaching team use only

truths <- tryCatch(
    {
        read.csv("../classification_test_label.csv")
    },
    error = function(e){
        read.csv("classification_test_label.csv")
    }
)

f1_score <- F1_Score(truths$x, pred.label)
cat(paste("f1_score is", f1_score))